src/crates.rs

```diff
@@ -0,0 +1,277 @@
+/// to a temporary folder. The temporary file and folder will be dropped
+/// once this goes out of scope.
+/// This retains a reference to the Config it was downloaded from, so it
+/// is tied to it's lifetime.
+pub struct CrateFile<'a> {
+    crate_file_prefix: String,
+    crate_file_extension: &'static str,
+    tempdir_handle: tempfile::TempDir,
+    config: &'a Config,
+}
+
+/// Represents a handle to a .crate file that has been downloaded to disk
+/// to a temporary folder, extracted, and built in release mode.
+/// The temporary file(s) and folder will be dropped once this goes out of scope.
+/// This retains a reference to the Config it was built from, so it
+/// is tied to it's lifetime.
+pub struct BuiltCrate<'a> {
+    crate_file_prefix: String,
+    tempdir_handle: tempfile::TempDir,
+    musl: bool,
+    config: &'a Config,
+}
+
+impl<'a> CrateFile<'a> {
+    /// Download the `.crate` file from `crates.io` to a temporary directory,
+    /// and if succesful, return a handle to it.
+    pub fn download_new(config: &Config) -> Result<CrateFile, Error> {
+        let pkgname = &config.package.name;
+        let pkgver = &config.package.version;
+        // This is downloaded to a temporary directory, instead of the current
+        // directory, to avoid the issue where cargo thinks its in a workspace.
+        // https://github.com/rust-lang/cargo/issues/5418
+        let tempdir_handle = tempfile::tempdir()?;
+        let crate_file_prefix = format!("{pkgname}-{pkgver}");
+        let crate_file_extension = "tar.gz";
+        let crate_file_path_full = tempdir_handle
+            .as_ref()
+            .join(&crate_file_prefix)
+            .with_extension(crate_file_extension);
+        let crate_url =
+            format!("https://static.crates.io/crates/{pkgname}/{pkgname}-{pkgver}.crate");
+        let success = Command::new("curl")
+            .arg("--output")
+            .arg(crate_file_path_full)
+            .arg(&crate_url)
+            .status()?
+            .success();
+        match success {
+            true => Ok(CrateFile {
+                tempdir_handle,
+                crate_file_prefix,
+                crate_file_extension,
+                config,
+            }),
+            false => Err(Error::DownloadingCrate { crate_url }),
+        }
+    }
+    /// Get the sha256sum for the downloaded .crate file.
+    // NOTE: Possibly future refactor target is crate::sha256sum
+    pub fn get_sha256sum(&self) -> Result<String, Error> {
+        let crate_file_path = self
+            .tempdir_handle
+            .as_ref()
+            .join(&self.crate_file_prefix)
+            .with_extension(self.crate_file_extension);
+        let bytes = std::fs::read(crate_file_path)?;
+        let digest = Hash::hash(&bytes);
+        let hex = digest.iter().fold(String::new(), |mut output, b| {
+            write!(output, "{:02x}", b).expect("Write to a string should not fail");
+            output
+        });
```

Comment:
Can you explain how this produces a sha256 that `makepkg` will agree to?

Reply:
This is the same method you used in the `sha256sum` function, and I applied this clippy lint: https://rust-lang.github.io/rust-clippy/master/index.html#/format_collect.

This can be verified by running `cargo-aur --source crates-op --no-bin`, and observing that the sha256sums in `target/cargo-sur/PKGBUILD` don't change after running `updpkgsums`.

Reply:
I guess past-me must have understood something about SHA that I've since forgotten.